src/NHibernate/Criterion/EntityProjection.cs

```diff
@@ -0,0 +1,181 @@
+		/// <param name="entityAlias">Entity alias</param>
+		public EntityProjection(System.Type rootEntity, string entityAlias)
+		{
+			_rootEntity = rootEntity;
+			_entityAlias = entityAlias;
+		}
+
+		/// <summary>
+		/// Fetch lazy properties
+		/// </summary>
+		public bool FetchLazyProperties { get; set; }
+
+		/// <summary>
+		/// Lazy load entity
+		/// </summary>
+		public bool Lazy { get; set; }
+
+		internal IQueryable Persister { get; private set; }
+		internal string ColumnAliasSuffix { get; private set; }
+		internal string TableAlias { get; private set; }
+
+		#region Configuration methods
+
+		/// <summary>
+		/// Lazy load entity
+		/// </summary>
+		public EntityProjection SetLazy(bool lazy = true)
+		{
+			Lazy = lazy;
+			return this;
+		}
+
+		/// <summary>
+		/// Fetch lazy properties
+		/// </summary>
+		public EntityProjection SetFetchLazyProperties(bool fetchLazyProperties = true)
+		{
+			FetchLazyProperties = fetchLazyProperties;
+			return this;
+		}
+
+		#endregion Configuration methods
+
+		#region IProjection implementation
+
+		string[] IProjection.Aliases => new[] { _entityAlias };
+
+		bool IProjection.IsAggregate => false;
+
+		bool IProjection.IsGrouped => false;
+
+		IType[] IProjection.GetTypes(string alias, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			return new[] { criteriaQuery.GetType(criteria, alias) };
+		}
+
+		IType[] IProjection.GetTypes(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _types;
+		}
+
+		string[] IProjection.GetColumnAliases(int position, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _identifierColumnAliases;
+		}
+
+		string[] IProjection.GetColumnAliases(string alias, int position, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _identifierColumnAliases;
+		}
+
+		SqlString IProjection.ToSqlString(ICriteria criteria, int position, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			string identifierSelectFragment = Persister.IdentifierSelectFragment(TableAlias, ColumnAliasSuffix);
+			return new SqlString(
+				Lazy
+					? identifierSelectFragment
+					: string.Concat(
+						identifierSelectFragment,
+						Persister.PropertySelectFragment(TableAlias, ColumnAliasSuffix, FetchLazyProperties)));
+		}
+
+		SqlString IProjection.ToGroupSqlString(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			throw new InvalidOperationException("not a grouping projection");
+		}
+
+		TypedValue[] IProjection.GetTypedValues(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			return Array.Empty<TypedValue>();
+		}
+
+		#endregion IProjection implementation
+
+		private void SetFields(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			//Persister is required, so let's use it as "initialized marker"
+			if (Persister != null)
```

Comment:
@fredericDelaporte Are you sure? From what I see `Subcriteria.SetProjection` use `rootCriteria`. Or did you mean some other scenario?

Reply:
Many methods of `EntityProjection` are calling `SetFields`, and these many methods can each be called with something else than the root criteria. Reading from the code, I am unable to ascertain the first `SetFields` call would forcibly be done with the root criteria.
Can you tell me if you are sure its first call can only be done with the root criteria, and why are you sure of this?

I do not get by the way why you refer to `Subcriteria.SetProjection`. Yes it delegates its implementation to the root criteria. But here I am pointing to `criteria.Alias`, which does not.

Reply:
My understanding is the following: from one side projections are assigned to some criterias. And from the other side projections methods should be called in context of assigned criteria (what the point of calling it for anything else?).
So it's more logical to me to use exact provided criteria for fields initialization.

>while it could receive a sub-criteria instead
>I do not get by the way why you refer to Subcriteria.SetProjection

And so if subcriteria projections are delegated to root criteria - I expect projections methods to be called in context of root criteria.

All 3 methods (`ToSqlString`, `GetTypes`, `GetColumnAliases`) entry points where SetFields is currently called are executed for root criteria:
https://github.com/nhibernate/nhibernate-core/blob/c020a4091368697807ede6154f10fd4a77bbe386/src/NHibernate/Loader/Criteria/CriteriaQueryTranslator.cs#L167-L187

Maybe I'm wrong but it looks like unnecessary complication with additional interface extension.

Reply:
All this entry points are public and may be called by code outside of NHibernate, so even if it were true they were always called with rootCriteria, there is still no guarantee about it. So if we agree `alias` must be taken from the root criteria when initializing a projection for root entity (for consistency with `GetRootEntityTypeIfAvailable` by the way, which always takes it from root criteria), then we should ensure that we will forcibly take it from the root criteria, even if something else than the root criteria is used to call one of these entry points.

Moreover, in current NHibernate code base, `IProjection.GetTypes` is also called from `InExpression.ToSqlString` (which is from `ICriterion`, not from `IProjection`), which can be called with criteria that do not look to be the root one. Same for `IProjection.ToSqlString`, called from `IdentifierEqExpression.ToSqlString`. Granted, it is unlikely a valid query could be done we those expressions calling an entity projection. But still, better be safe in my opinion.

The `rootCriteria.ProjectionCriteria` sent in `IProjection.ToSqlString` adds also to the risk. Currently it seems always `null` or being equal to `rootCriteria`, being affected only by `this`. (I guess it will not be `null` if that call occurs by the way, otherwise we are in troubles too). But some code (the `Clone`) assumes it could be something else too. Maybe that is an obsolete thing, or maybe that is here for being future proof.

The contract of those entry points does not stipulate they must be called with the root criteria, so relying on this is not future proof.